crates/precompile/src/kzg_point_evaluation.rs

```diff
@@ -86,7 +96,122 @@ pub fn verify_kzg_proof(commitment: &Bytes48, z: &Bytes32, y: &Bytes32, proof: &
             let env = kzg_rs::EnvKzgSettings::default();
             let kzg_settings = env.get();
             KzgProof::verify_kzg_proof(commitment, z, y, proof, kzg_settings).unwrap_or(false)
+        } else {
+            bls12_381_backend::verify_kzg_proof(commitment, z, y, proof)
+        }
+    }
+}
+
+/// BLS12-381 backend implementation for KZG verification
+#[cfg(not(any(feature = "c-kzg", feature = "kzg-rs")))]
+mod bls12_381_backend {
+    use super::*;
+    use ark_bls12_381::{Bls12_381, Fr, G1Affine, G2Affine};
+    use ark_ec::{pairing::Pairing, AffineRepr, CurveGroup};
+    use ark_ff::{BigInteger, One, PrimeField};
+    use ark_serialize::CanonicalDeserialize;
+    use core::ops::Neg;
+    use std::string::ToString;
+
+    /// Verify KZG proof using BLS12-381 implementation.
+    ///
+    /// https://github.com/ethereum/EIPs/blob/4d2a00692bb131366ede1a16eced2b0e25b1bf99/EIPS/eip-4844.md?plain=1#L203
+    /// https://github.com/ethereum/consensus-specs/blob/master/specs/deneb/polynomial-commitments.md#verify_kzg_proof_impl
+    #[inline]
+    pub(super) fn verify_kzg_proof(
+        commitment: &Bytes48,
+        z: &Bytes32,
+        y: &Bytes32,
+        proof: &Bytes48,
+    ) -> bool {
+        // Parse the commitment (G1 point)
+        let Ok(commitment_point) = parse_g1_compressed(&commitment.0) else {
+            return false;
+        };
+
+        // Parse the proof (G1 point)
+        let Ok(proof_point) = parse_g1_compressed(&proof.0) else {
+            return false;
+        };
+
+        // Parse z and y as field elements (Fr, scalar field)
+        // We expect 32-byte big-endian scalars that must be canonical
+        let Ok(z_fr) = read_scalar_canonical(&z.0) else {
+            return false;
+        };
+        let Ok(y_fr) = read_scalar_canonical(&y.0) else {
+            return false;
+        };
+
+        // Get the trusted setup G2 point [τ]₂
+        // TODO: This only needs to be done once and cached
+        let tau_g2 = get_trusted_setup_g2();
```

Comment:
This should be cached. 

I did some rough benchmarks of this and it was taking about 60 microseconds, so its not taking up a lot of time.

 The whole function takes 1ms for arkworks and about 0.8ms for c-kzg --the 20% is likely coming from the pairing being more efficient, this could be checked by using blst instead of arkworks since c-kzg using blst